gadopt/scalar_equation.py

```diff
@@ -93,165 +74,92 @@ class ScalarDiffusionTerm(BaseTerm):
-            F += -inner(avg(dot(diff_tensor, grad(phi))), jump(q, n))*self.dS
-            F += -inner(jump(phi, n), avg(dot(diff_tensor, grad(q))))*self.dS
-
-        for id, bc in bcs.items():
-            if 'q' in bc:
-                jump_q = trial-bc['q']
-                sigma_ext = sigma * fd.FacetArea(self.mesh)/fd.CellVolume(self.mesh)
-                # this corresponds to the same 3 terms as the dS integrals for DG above:
-                F += 2*sigma_ext*phi*inner(n, dot(diff_tensor, n))*jump_q*self.ds(id)
-                F += -inner(dot(diff_tensor, grad(phi)), n)*jump_q*self.ds(id)
-                F += -inner(phi*n, dot(diff_tensor, grad(q))) * self.ds(id)
-                if 'flux' in bc:
-                    raise ValueError("Cannot apply both `q` and `flux` bc on same boundary")
-            elif 'flux' in bc:
-                # here we need only the third term, because we assume jump_q=0 (q_ext=q)
-                # the provided flux = kappa dq/dn = dot(n, dot(diff_tensor, grad(q))
-                F += -phi*bc['flux']*self.ds(id)
-
-        return -F
-
-
-class ScalarSourceTerm(BaseTerm):
+
+    F = inner(grad(eq.test), dot(diff_tensor, grad(q))) * eq.dx
+
+    sigma = interior_penalty_factor(eq, shift=-1)
+    if not is_continuous(eq.trial_space):
+        sigma_int = sigma * avg(FacetArea(eq.mesh) / CellVolume(eq.mesh))
+        F += (
+            sigma_int
+            * inner(jump(eq.test, eq.n), dot(avg(diff_tensor), jump(q, eq.n)))
+            * eq.dS
+        )
+        F -= inner(avg(dot(diff_tensor, grad(eq.test))), jump(q, eq.n)) * eq.dS
+        F -= inner(jump(eq.test, eq.n), avg(dot(diff_tensor, grad(q)))) * eq.dS
+
+    for bc_id, bc in eq.bcs.items():
+        if "q" in bc and "flux" in bc:
+            raise ValueError("Cannot apply both `q` and `flux` on the same boundary.")
+
+        if "q" in bc:
+            jump_q = trial - bc["q"]
+            sigma_ext = sigma * FacetArea(eq.mesh) / CellVolume(eq.mesh)
+            # Terms below are similar to the above terms for the DG dS integrals.
+            F += (
+                2
+                * sigma_ext
+                * eq.test
+                * inner(eq.n, dot(diff_tensor, eq.n))
+                * jump_q
+                * eq.ds(bc_id)
+            )
+            F -= inner(dot(diff_tensor, grad(eq.test)), eq.n) * jump_q * eq.ds(bc_id)
+            F -= inner(eq.test * eq.n, dot(diff_tensor, grad(q))) * eq.ds(bc_id)
+        elif "flux" in bc:
+            # Here we need only the third term because we assume jump_q = 0
+            # (q_ext = trial) and flux = kappa dq/dn = dot(n, dot(diff_tensor, grad(q)).
+            F -= eq.test * bc["flux"] * eq.ds(bc_id)
+
+    return -F
+
+
+def scalar_source_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
     r"""Scalar source term `s_T`."""
+    source = eq.source if hasattr(eq, "source") else 0
```

Comment:
I think the previous early return behaviour is preferable to always computing a dot product with 0.